### PR TITLE
feat: constant folding implemented for core and float extension

### LINKED ISSUE #711: Constant folding pass
Some kind of way of constant folding over extension constants.

### PR DESCRIPTION
**Pull request by ss2165**
Closes #711 
CI will pass after updating MSRV to 1.75 (from end of year)

based on #757

### CODE REVIEW

File: src/std_extensions/arithmetic/int_ops.rs [hunk 1]
```diff
@@ -11,6 +11,7 @@ use crate::ops::OpName;
 use crate::type_row;
 use crate::types::{FunctionType, PolyFuncType};
 use crate::utils::collect_array;
+
```
Comment (lmondada) [inline src/std_extensions/arithmetic/int_ops.rs:14]: Out of curiosity: are you following some style guide regarding newlines here? Would like to know for my own coding if there is a recommended style.

File: src/std_extensions/arithmetic/float_ops/fold.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
+use crate::{
```
Comment (lmondada) [inline src/std_extensions/arithmetic/float_ops/fold.rs:1]: Consider naming this file `const_fold.rs` to match the extension name?

File: src/extension/op_def.rs [hunk 3]
```diff
@@ -307,6 +308,9 @@ pub struct OpDef {
     // can only treat them as opaque/black-box ops.
     #[serde(flatten)]
     lower_funcs: Vec<LowerFunc>,
+
+    #[serde(skip)]
+    constant_folder: Box<dyn ConstFold>,
```
Comment (lmondada) [inline src/extension/op_def.rs:313]: Can you add a short docstring here?

File: src/algorithm/const_fold.rs [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+            if let Value::Tuple { vs } = c.value() {
+                if let TypeEnum::Tuple(tys) = c.const_type().as_type_enum() {
+                    return out_row(tys.iter().zip(vs.iter()).map(|(t, v)| {
+                        Const::new(v.clone(), t.clone())
+                            .expect("types should already have been checked")
+                    }));
+                }
+            }
+            None // could panic
```
Comment (lmondada) [inline src/algorithm/const_fold.rs:66]: I am not sure the comment is correct, i.e. what would `panic` and how returning None would help. I suppose you mean that the `let if` clauses might fail?

File: src/algorithm/const_fold.rs [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+            break;
+        }
+        for (replace, removes) in rewrites {
+            h.apply_rewrite(replace).unwrap();
+            for rem in removes {
+                if let Ok(const_node) = h.apply_rewrite(rem) {
+                    // if the LoadConst was removed, try removing the Const too.
+                    if h.apply_rewrite(RemoveConst(const_node)).is_err() {
+                        // const cannot be removed - no problem
+                        continue;
```
Comment (lmondada) [inline src/algorithm/const_fold.rs:208]: I find it surprising that this is not removed by the rewrite itself. Maybe add a comment about this in the docstring of `find_consts`?

**Review by lmondada — APPROVED** [review]
Thanks Seyon, this will be super useful!

File: src/algorithm/const_fold.rs (hunk 4 above)
Comment (ss2165) [reply src/algorithm/const_fold.rs:66]: i mean that `None` shouldn't ever be returned here, so it might be clearer to panic

**Pull request merged**